plutus-tx/README.md

```diff
@@ -1,3 +1,228 @@
+
+### Using PlutusTx builtins
+
+PlutusTx has some builtin types and functions available, both for working with primitive
+data (integers and bytestrings), and also for performing chain-specific operations.
+
+The PlutusTx builtins are available via the `Language.PlutusTx.Builtins` module. You
+shouldn't need to use the integer and bytestring builtins directly, these are mapped
+to the corresponding Haskell operations on `Int` and lazy `ByteString` directly. However,
+you may wish to use some of the others.
+
+`error` deserves a special mention. `error` causes the transaction to abort when it is
+evaluated, which is the way that validation failure is signalled.
+
+### Lifting values
+
+This is all very good for defining pieces of code *statically*, but you are
+likely to want to *dynamically* produce Plutus Core programs. For example, you
+might be writing the body of a transaction to initiate a crowdfunding smart contract,
+which would need to be parameterized by user input determining the size of the goal,
+the campaign start and end times, etc.
+
+You can do this by writing the static code as a function, and then passing an
+argument at runtime by *lifting* it and then applying the two programs together. As a
+very simple example, let's write an add-one function.
+
+```haskell
+-- TODO: show PIR here too
+
+-- |
+-- >>> let program = addOneToN 4
+-- >>> prettyPlcDef program
+-- (program 1.0.0
+--   [
+--     [
+--       (lam
+--         addInteger
+--         (fun [(con integer) (con 8)] (fun [(con integer) (con 8)] [(con integer) (con 8)]))
+--         (lam ds [(con integer) (con 8)] [ [ addInteger ds ] (con 8 ! 1) ])
+--       )
+--       { (builtin addInteger) (con 8) }
+--     ]
+--     (con 8 ! 4)
+--   ]
+-- )
+
+-- >>> prettyPlcDef $ runCk program
+-- (con 8 ! 5)
+addOneToN :: Int -> Program TyName Name ()
+addOneToN n =
+    let
+        addOne = $$(plutus [|| \(x:: Int) -> x + 1 ||])
+        -- TODO: This looks terrible
+        arg = Program () (defaultVersion ()) (runQuote $ unsafeLiftPlc n)
+    in applyProgram (getAst addOne) arg
+```
+
+Here we have lifted the Haskell value `4` into a Plutus Core term at runtime.
+In order to do this, a type must have an instance of the `LiftPir` class. In
+practice, you should generate these with the `makeLift` TH function from
+`Laguage.PlutusTx.Lift`. This makes it easy to use the same types both inside your
+PlutusTx program and in the external code that uses it.
+
+The combined program just applies the compiled lambda to the lifted value
+(notice that the lambda is a bit complicated now since we have compiled the addition
+into a builtin). We've then used the CK evaluator for Plutus Core to evaluate
+the program and check that the result was what we expected
+
+This is more or less all you need to get started - from here on it should mostly
+be like writing normal Haskell.
+
+```haskell
+main :: IO ()
+main = do
+    _ <- traverse (putStrLn . show . prettyPlcDef . getAst) [
```

Comment:
What is this for? (Presumably for debugging but there should be a comment)

Reply:
It's so that the file can be a test suite so it gets compiled. But thinking about it, that's unnecessary since doctest will interpret everything so will fail if it doesn't compile.